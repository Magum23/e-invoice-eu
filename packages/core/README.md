<p align="center">
	<img
		src="../../assets/e-invoice-eu-logo-2.webp"
		width="256" height="256" />
</p>

[![licence](https://img.shields.io/badge/licence-WTFPL-blue)](http://www.wtfpl.net/)
[![price](https://img.shields.io/badge/price-FREE-green)](https://github.com/gflohr/qgoda/blob/main/LICENSE)
[![coverage](https://img.shields.io/coverallsCoverage/github/gflohr/e-invoice-eu?branch=main)](https://coveralls.io/github/gflohr/e-invoice-eu?branch=main)
[![documentation](https://img.shields.io/badge/documentation-Qgodaüçì-ffc107)](https://www.qgoda.net/)
[![stand with](https://img.shields.io/badge/stand%20with-Ukraineüá∫üá¶-ffc107)](https://www.standwithukraineeurope.com/en//)

# E-Invoice-EU<!-- omit from toc -->

Free and open source tool chain for generating EN16931 conforming e-invoices
(Factur-X/ZUGFeRD, UBL, CII, XRechnung) from popular spreadsheet formats or
JSON.

This package contains the core library of `@e-invoice-eu/core`. Both the
[E-Invoice-EU command line tool](https://github.com/gflohr/e-invoice-eu/tree/main/apps/cli)
and the [E-Invoice-EU web service](https://github.com/gflohr/e-invoice-eu/tree/main/apps/server)
use it as their foundation.

## Table of Contents<!-- omit from toc -->

- [Features](#features)
- [Documentation](#documentation)
- [Installation](#installation)
- [Usage](#usage)
	- [Importing/Requiring the Library](#importingrequiring-the-library)
	- [Creating Invoice Data from Spreadsheet Files](#creating-invoice-data-from-spreadsheet-files)
	- [Generating an Invoice from Data in the Internal Format](#generating-an-invoice-from-data-in-the-internal-format)
		- [Argument `format`](#argument-format)
		- [Argument `lang`](#argument-lang)
		- [Argument `embedPDF`](#argument-embedpdf)
		- [Argument `pdf`](#argument-pdf)
		- [Argument `data`](#argument-data)
		- [Argument `attachments`](#argument-attachments)
	- [Getting Supported Formats](#getting-supported-formats)
	- [Getting the JSON Schema Definitions](#getting-the-json-schema-definitions)
- [Copyright](#copyright)
- [Disclaimer](#disclaimer)

## Features

The library can create e-invoices in the following formats:

- Cross Industry Invoice - CII
- Universal Business Language - UBL
- Factur-X/ZUGFeRD (all profiles including XRECHNUNG) with full PDF/A support

E-Invoices can be created directly from data in the [E-Invoice-EU internal
format](https://gflohr.github.io/e-invoice-eu/en/docs/details/internal-format/)
or via a [mapping definition](https://gflohr.github.io/e-invoice-eu/en/docs/details/mapping/)
from popular spreadsheet formats like `.ods`, `.xlsx`, `.csv`, ...

The library is implemented in TypeScript and provides complete type definitions
out of the box. No additional @types package is required. All data structures,
including those for the internal invoice format and mapping definitions, are
fully typed, ensuring comprehensive IntelliSense support.

The library is available as an ES module, CommonJS, and UMD, making it
compatible with both browser environments and JavaScript runtimes like Node.js.

## Documentation

The general documentation of `e-invoice-eu` is available at the
[`e-invoice-eu` GitHub page](https://github.com/gflohr/e-invoice-eu/tree/main/packages/core).
For reference, you can also consult the
[API documentation](https://gflohr.github.io/e-invoice-eu/api-docs/).

## Installation

```sh
npm install @e-invoice-eu/core
```

## Usage

### Importing/Requiring the Library

If you have the `import` keyword:

```typescript
import { InvoiceService, MappingService } from '@e-invoice-eu/core';
```

With `require`:

```javascript
const { InvoiceService, MappingService } = require('@e-invoice-eu/core');
```

In the browser:

```html
<script src="https://cdn.jsdelivr.net/npm/@e-invoice-eu/core/dist/esgettext.min.js"></script>
<script>
	const invoiceService = new eInvoiceEU.InvoiceService(/* arguments */);
	const mappingService = new eInvoiceEU.MappingService(/* arguments */);
	// Your code goes here ...
</script>
```

### Creating Invoice Data from Spreadsheet Files

For this optional step, you need both the spreadsheet file as a `Buffer` and a
mapping definition as a
[`Mapping`](https://gflohr.github.io/e-invoice-eu/api-docs/interfaces/Mapping.html)
object:

```typescript
import { MappingService } from '@e-invoice-eu/core';

const mappingService = new MappingService(console);
const invoice = mappingService.transform(
	spreadsheet, // a `Buffer`
	'Factur-X-Extended', // see https://gflohr.github.io/e-invoice-eu/en/docs/basics/supported-formats/
	mapping, // a `Mapping`
);
```

The returned `invoice` object is an
[`Invoice`](https://gflohr.github.io/e-invoice-eu/api-docs/interfaces/Invoice.html)
instance in the internal format. It is not yet an XML string or PDF buffer!

### Generating an Invoice from Data in the Internal Format

You can either provide the invoice data yourself as an
[`Invoice`](https://gflohr.github.io/e-invoice-eu/api-docs/interfaces/Invoice.html)
instance or you can create it from spreadsheet data and a mapping definition
as shown in the [previous section](#creating-invoice-data-from-spreadsheet-files).

```typescript
import { InvoiceService } from '@e-invoice-eu/core';

const invoiceService = new InvoiceService(console);

const renderedInvoice = await invoiceService.generate(invoiceData, options);
```

See this summary for the `options` (optional options are marked with an
question mark `?`):

| Name         | Type                                                                               | Description                                                                                   |
| ------------ | ---------------------------------------------------------------------------------- | --------------------------------------------------------------------------------------------- |
| format       | `string`                                                                           | a [supported format](https://gflohr.github.io/e-invoice-eu/en/docs/basics/supported-formats/) |
| lang         | `string`                                                                           | a language identifier like `fr-fr`                                                            |
| embedPDF?    | `boolean`                                                                          | pass `true` if a PDF version should be embedded into XML output                               |
| pdf?         | [`FileInfo`](https://gflohr.github.io/e-invoice-eu/api-docs/types/FileInfo.html)   | the PDF version of the invoice                                                                |
| data?        | [`FileInfo`](https://gflohr.github.io/e-invoice-eu/api-docs/types/FileInfo.html)   | invoice spreadsheet file                                                                      |
| attachments? | [`FileInfo[]`](https://gflohr.github.io/e-invoice-eu/api-docs/types/FileInfo.html) | an arbitrary number of attachments                                                            |

#### Argument `format`

This contains the format of the e-invoice as a case-insensitive string. You
can also use aliases. For example, `Factur-X-Comfort` is an alias for
`Factur-X-EN16931`.

#### Argument `lang`

A language identifier like `fr-fr`. This is only used for the Factur-X/ZUGFeRD
formats for some canned texts in the PDF XMP meta data.

#### Argument `embedPDF`

This is only used for the pure XML formats (everything that is not
Factur-X/ZUGFeRD). If it has a truthy value, a PDF version of the invoice
is embedded as a base64 encoded string in the XML.

#### Argument `pdf`

This is required in two cases:

1. The format is a Factur-X/ZUGFeRD format and `data` was **not** specified.
2. The format is a pure XML format, `embedPDF` was specified and `data` was not specified.

#### Argument `data`

A spreadsheet version of the invoice. If a PDF is required, either because
embedPDF was specified for pure XML or because the format is Factur-X/ZUGFeRD,
[LibreOffice](https://libreoffice.org) is invoked for rendering the spreadsheet
as a PDF.

This feature will trigger an exception if used in the browser.

#### Argument `attachments`

You can specify an arbitrary number of additional attachments.

In case of the pure XML formats, they are embedded as base64 encoded strings
in the XML. In case of Factur-X/ZUGFeRD they are attached to the PDF,
additionally to the mandatory attachment `factur-x.xml`.

### Getting Supported Formats

```typescript
import { FormatFactoryService } from '@e-invoice-eu/core';

const factoryService = new FormatFactoryService();
const formats = factoryService.listFormatServices();
```

In `format` you will find an array of
[`FormatInfo`](https://gflohr.github.io/e-invoice-eu/api-docs/types/FormatInfo.html)
objects.

### Getting the JSON Schema Definitions

```typescript
import { invoiceSchema, mappingSchema } from '@e-invoice-eu/core';
```

These variables contain the schema for the
[`Invoice`](https://gflohr.github.io/e-invoice-eu/api-docs/types/Invoice.html)
and [`Mapping`](https://gflohr.github.io/e-invoice-eu/api-docs/types/Mapping.html)
interfaces. These schemas can be passed as an argument to the compile method of an Ajv instance, see https://ajv.js.org/api.html#ajv-compile-schema-object-data-any-boolean-promise-any for more information!

The schemas have the type
[`JSONSchemaType<Invoice>`](https://ajv.js.org/guide/typescript.html#utility-types-for-schemas)
and
[`JSONSchemaType<Mapping>`](https://ajv.js.org/guide/typescript.html#utility-types-for-schemas)
respectively.

## Copyright

Copyright (C) 2024-2025 Guido Flohr <guido.flohr@cantanea.com>, all
rights reserved.

This is free software available under the terms of the
[WTFPL](http://www.wtfpl.net/).

## Disclaimer

This free software has been written with the greatest possible care, but like
all software it may contain errors. Use at your own risk! There is no
warranty and no liability.
