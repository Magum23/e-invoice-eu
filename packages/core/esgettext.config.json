{
	"package": {
		"name": "e-invoice-eu",
		"version": "1.0.0",
		"textdomain": "e-invoice-eu",
		"msgid-bugs-address": "https://github.com/gflohr/e-invoice-eu/issues"
	},
	"po": {
		"directory": "po",
		"locales": ["de"]
	},
	"install": {
		"directory": "src/locale"
	},
	"programs": {
		"msgmerge": {
			"options": ["previous"]
		},
		"msgfmt": {
			"options": ["check", "statistics", "verbose"]
		}
	}
}
