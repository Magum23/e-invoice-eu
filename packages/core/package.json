{
	"name": "@e-invoice-eu/core",
	"version": "1.4.2",
	"description": "Generate e-invoices (E-Rechnung in German) conforming to EN16931 (Factur-X/ZUGFeRD, UBL, CII, XRechnung aka X-Rechnung) from LibreOffice Calc/Excel data or JSON. ",
	"keywords": [
		"zugferd",
		"e-invoice",
		"ubl",
		"pdfa",
		"cii",
		"en16931",
		"xrechnung",
		"factur-x",
		"e-rechnung"
	],
	"main": "dist/e-invoice-eu.cjs.js",
	"module": "dist/e-invoice-eu.esm.js",
	"typings": "dist/index.d.ts",
	"publishConfig": {
		"access": "public"
	},
	"repository": {
		"type": "git",
		"url": "git+https://github.com/gflohr/e-invoice-eu.git"
	},
	"bugs": {
		"url": "https://github.com/gflohr/e-invoice-eu/issues"
	},
	"homepage": "https://github.com/gflohr/e-invoice-eu/packages/runtime",
	"scripts": {
		"build": "rimraf dist && nest build",
		"prepublishOnly": "npm run build",
		"check:clean": "../../check-clean",
		"format": "prettier --write *.json 'src/**/*.ts'",
		"watch": "tsc --watch",
		"lint": "eslint './src/**/*.ts'",
		"lint:fix": "eslint --fix './src/**/*.ts'",
		"lint:fix-dry-run": "eslint --fix-dry-run './src/**/*.ts'",
		"lint:staged": "lint-staged",
		"typedoc": "typedoc --out ../docs/api-docs --excludeNotDocumented --excludePrivate --excludeProtected src",
		"test": "jest",
		"test:watch": "jest --watch",
		"test:cov": "jest --coverage",
		"test:debug": "node --inspect-brk -r tsconfig-paths/register -r ts-node/register node_modules/.bin/jest --runInBand"
	},
	"author": "Guido Flohr <guido.flohr@cantanea.com> (http://www.guido.flohr.net/)",
	"license": "WTFPL",
	"files": [
		"dist"
	],
	"jest": {
		"moduleFileExtensions": [
			"js",
			"json",
			"ts"
		],
		"rootDir": "src",
		"testRegex": ".*\\.spec\\.ts$",
		"transform": {
			"^.+\\.(t|j)s$": "ts-jest"
		},
		"collectCoverageFrom": [
			"**/*.(t|j)s"
		],
		"coverageDirectory": "../coverage",
		"testEnvironment": "node"
	},
	"lint-staged": {
		"*.{js,ts,json,scss,css,md}": [
			"prettier --write"
		],
		"src/**/*.ts": [
			"eslint"
		]
	},
	"devDependencies": {
		"@rollup/plugin-typescript": "^12.1.1",
		"eslint": "^9.16.0",
		"jest": "^29.7.0",
		"lint-staged": "^15.2.10",
		"rimraf": "^6.0.1",
		"rollup": "^4.28.0",
		"ts-jest": "^29.2.5",
		"ts-loader": "^9.5.1",
		"typedoc": "^0.27.9",
		"typescript": "^5.7.2",
		"typescript-eslint": "^8.17.0",
		"webpack-cli": "^5.1.4",
		"xhr-mock": "^2.5.1"
	},
	"dependencies": {
		"ajv": "^8.17.1"
	}
}
