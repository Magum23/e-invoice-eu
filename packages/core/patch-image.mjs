import * as fs from 'fs/promises';

async function patchImage() {
	const filename = '../docs/api-docs/index.html';
	const readme = await fs.readFile('../docs/api-docs/index.html', 'utf-8');
	const patched = readme.replace(
		'src="../../assets/e-invoice-eu-logo-2.webp"',
		'src="https://raw.githubusercontent.com/gflohr/e-invoice-eu/main/assets/e-invoice-eu-logo-2.webp"',
	);
	await fs.writeFile(filename, patched, 'utf-8');
}

patchImage();
