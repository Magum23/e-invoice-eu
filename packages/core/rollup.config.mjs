import commonjs from '@rollup/plugin-commonjs';
import json from '@rollup/plugin-json';
import resolve from '@rollup/plugin-node-resolve';
import replace from '@rollup/plugin-replace';
import terser from '@rollup/plugin-terser';
import typescript from '@rollup/plugin-typescript';
import * as fs from 'fs';

const pkg = JSON.parse(
	fs.readFileSync('./package.json', { encoding: 'utf-8' }),
);

export default [
	// UMD builds for the browser.
	{
		input: 'src/index.ts',
		output: {
			name: 'eInvoiceEU',
			file: './dist/e-invoice-eu.min.js',
			format: 'umd',
			sourcemap: true,
		},
		plugins: [
			// Replace the import for the PDF creation.
			replace({
				values: {
					'../utils/render-spreadsheet': JSON.stringify('../utils/render-spreadsheet.browser'),
				},
				delimiters: ["'", "'"],
				preventAssignment: true,
			}),
			json(),
			resolve({ preferBuiltins: true }),
			commonjs(),
			typescript({
				exclude: 'src/**/*.spec.ts',
			}),
			terser(),
		],
	},
	{
		input: 'src/index.ts',
		output: {
			name: 'esgettext',
			file: './dist/e-invoice-eu.js',
			format: 'umd',
			sourcemap: true,
		},
		plugins: [
			// Replace the import for the PDF creation.
			replace({
				values: {
					'../utils/render-spreadsheet': JSON.stringify('../utils/render-spreadsheet.browser'),
				},
				delimiters: ["'", "'"],
				preventAssignment: true,
			}),
			json(),
			resolve({ preferBuiltins: true }),
			commonjs(),
			typescript({
				exclude: 'src/**/*.spec.ts',
			}),
		],
	},
	{
		input: 'src/index.ts',
		external: ['fs'],
		plugins: [
			json(),
			typescript({
				exclude: 'src/**/*.spec.ts',
			}),
		],
		output: [
			{ file: pkg.main, format: 'cjs', sourcemap: true },
			{ file: pkg.module, format: 'es', sourcemap: true },
		],
	},
];
