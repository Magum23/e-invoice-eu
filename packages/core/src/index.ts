/* istanbul ignore file */
export * from './invoice';
export * from './mapping';
export * from './validation';
