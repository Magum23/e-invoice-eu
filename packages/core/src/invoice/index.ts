/* istanbul ignore file */
export { Invoice } from './invoice.interface';
export * from './invoice.schema';
export * from './invoice.service';
