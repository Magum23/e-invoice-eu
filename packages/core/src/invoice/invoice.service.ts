import Ajv2019, { ValidateFunction } from 'ajv/dist/2019';

import { Invoice } from './invoice.interface';
import { invoiceSchema } from './invoice.schema';
import { FormatFactoryService } from '../format/format.factory.service';
import { ILogger } from '../ilogger';
import { ValidationService } from '../validation';

/**
 * A container for a file used for an e-invoice.
 */
export type InvoiceFile = {
	/**
	 * The data of the file.
	 */
	buffer: Buffer;

	/**
	 * The filename.
	 */
	filename: string;

	/**
	 * The MIME type of the file.
	 */
	mimetype: string;
};

export type InvoiceAttachment = {
	/**
	 * The uploaded file.
	 */
	file: InvoiceFile;

	/**
	 * An optional ID.
	 */
	id?: string;

	/**
	 * An optional description.
	 */
	description?: string;
};

export type InvoiceServiceOptions = {
	/**
	 * The invoice format like `XRECHNUNG-UBL` or `Factur-X-Extended`.
	 */
	format: string;

	/**
	 * The spreadsheet data.
	 */
	data?: InvoiceFile;

	/**
	 * A PDF version of the invoice.  For Factur-X, either `data` or `pdf`
	 * must be present.
	 */
	pdf?: InvoiceFile;

	/**
	 * A language identifier like "fr-ca".
	 */
	lang: string;

	/**
	 * An array of supplementary attachments.
	 */
	attachments: InvoiceAttachment[];

	/**
	 * Set to invoice description if invoice should be embedded.
	 */
	embedPDF?: boolean;

	/**
	 * ID for an embedded PDF, defaults to the document id.
	 */
	pdfID?: string;

	/**
	 * Description for the embedded PDF.
	 */
	pdfDescription?: string;

	/**
	 * Path to LibreOffice executable.
	 */
	libreOfficePath?: string;
};

export class InvoiceService {
	private readonly formatFactoryService: FormatFactoryService;
	private readonly validator: ValidateFunction<Invoice>;
	private readonly validationService: ValidationService;

	constructor(
		private readonly logger: ILogger,
	) {
		this.formatFactoryService = new FormatFactoryService();
		const ajv = new Ajv2019({
			strict: true,
			allErrors: true,
			useDefaults: true,
		});
		this.validator = ajv.compile(invoiceSchema);
		this.validationService = new ValidationService(this.logger);
	}

	async generate(
		input: unknown,
		options: InvoiceServiceOptions,
	): Promise<string | Buffer> {
		const invoice = this.validationService.validate(
			'invoice data',
			this.validator,
			input,
		);

		const formatter = this.formatFactoryService.createFormatService(
			options.format,
		);

		if (typeof formatter === 'undefined') {
			throw new Error(`Unsupported format: ${options.format}`);
		}

		return formatter.generate(invoice, options);
	}
}
