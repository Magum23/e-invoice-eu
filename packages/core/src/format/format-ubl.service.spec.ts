import { Invoice, Mapping } from '@e-invoice-eu/core';

import { FormatUBLService } from './format-ubl.service';
import { InvoiceServiceOptions } from '../invoice/invoice.service';
import { Logger } from '../logger.interface';

describe('UBL', () => {
	let service: FormatUBLService;

	beforeEach(async () => {
		service = new FormatUBLService({} as unknown as Logger);
	});

	it('should be defined', () => {
		expect(service).toBeDefined();
	});

	it('should have a customization id', () => {
		expect(service.customizationID).toBeDefined();
	});

	it('should have a profile id', () => {
		expect(service.profileID).toBeDefined();
	});

	it('should fill in mapping defaults', () => {
		const mapping = { 'ubl:Invoice': {} } as unknown as Mapping;

		service.fillMappingDefaults(mapping);
		expect(mapping['ubl:Invoice']['cbc:CustomizationID']).toBe(
			service.customizationID,
		);
		expect(mapping['ubl:Invoice']['cbc:ProfileID']).toBe(service.profileID);
	});

	it('should fill in invoice defaults', () => {
		const invoice = { 'ubl:Invoice': {} } as unknown as Invoice;

		service.fillInvoiceDefaults(invoice);
		expect(invoice['ubl:Invoice']['cbc:CustomizationID']).toBe(
			service.customizationID,
		);
		expect(invoice['ubl:Invoice']['cbc:ProfileID']).toBe(service.profileID);
	});

	it('should generate XML', async () => {
		const invoice: Invoice = { 'ubl:Invoice': {} } as unknown as Invoice;
		const options = { attachments: [] } as unknown as InvoiceServiceOptions;

		const xml = await service.generate(invoice, options);

		expect(xml).toMatchSnapshot();
	});
});
