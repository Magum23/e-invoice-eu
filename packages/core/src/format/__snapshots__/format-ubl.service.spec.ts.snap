// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UBL should embed a PDF 1`] = `
"<?xml version="1.0" encoding="utf-8"?>
<Invoice xmlns="urn:oasis:names:specification:ubl:schema:xsd:Invoice-2" xmlns:cac="urn:oasis:names:specification:ubl:schema:xsd:CommonAggregateComponents-2" xmlns:cbc="urn:oasis:names:specification:ubl:schema:xsd:CommonBasicComponents-2">
	<cac:AdditionalDocumentReference>
		<cac:Attachment>
			<cbc:EmbeddedDocumentBinaryObject mimeCode="application/pdf" filename="invoice.pdf">SW52b2ljZQ==</cbc:EmbeddedDocumentBinaryObject>
		</cac:Attachment>
	</cac:AdditionalDocumentReference>
</Invoice>"
`;

exports[`UBL should generate XML 1`] = `
"<?xml version="1.0" encoding="utf-8"?>
<Invoice xmlns="urn:oasis:names:specification:ubl:schema:xsd:Invoice-2" xmlns:cac="urn:oasis:names:specification:ubl:schema:xsd:CommonAggregateComponents-2" xmlns:cbc="urn:oasis:names:specification:ubl:schema:xsd:CommonBasicComponents-2"/>"
`;
